p4-16/psa/PSA.mdk

```diff
@@ -663,13 +661,98 @@ updating of `Counter` and `DirectCounter` externs.
+or "marking" packets that exceed an average packet or bit rate.  To
+"mark" a packet means to change one or more of its quality of service
+values in packet headers such as the 802.1Q PCP (priority code point)
+or DSCP (differentiated service code point) bits within the IPv4 or
+IPv6 type of service byte.  The meters specified in the PSA are
+3-color meters.
+
+PSA meters do not require any particular drop or marking actions, nor
+do they automatically implement those behaviors for you.  Meters keep
+enough state, and update their state during `execute()` method calls,
+in such a way that they return a green (also known as conform), yellow
+(exceed), or red (violate) result.  See RFC 2698 for details on the
+conditions under which one of these three results is returned.  The P4
+program is responsible for examining that returned result, and making
+changes to packet forwarding behavior as a result.
+
+Similar to counters, there are two flavors of meters: indexed and
+direct.  (Indexed) meters are addressed by index, while direct meters
+are addressed using P4Runtime table entry as key.
+
+There are many other similarities between counters and meters,
+including:
+
+- The number of independently updatable meter values.
+- Where meter updates are allowed in a P4 program.
+- For `bytes` type meters, the packet length used in the update is
+  determined by the PSA implementation, and can vary from one PSA
+  implementation to another.
+
+Further similarities between direct counters and direct meters
+include:
+
+- `DirectMeter` `execute` method calls must be performed within
+  actions invoked by the table that owns the `DirectMeter` instance.
+  It is optional for such an action to call the `execute` method.
+- There must be a meter state associated with a `DirectMeter`
+  instance's owner table, that can be updated when the table result is
+  a miss.  As for a `DirectCounter`, this state only needs to exist if
+  a default action is assigned to the table.
+
+The table attribute to specify that a table owns a `DirectMeter`
+instance is `psa_direct_meters`.  The value of this table attribute is
+a list of meter instances.
+
+As for counters, if you call the `execute(idx)` method on an indexed
+meter and `idx` is at least the number of meter states, so `idx` is
+out of range, no meter state is updated.  The `execute` call still
+returns a value of type `MeterColor_t`, but the value is undefined --
+programs that wish to have predictable behavior across implementations
+must not use the undefined value in a way that affects the output
+packet or other side effects.  The example code below shows one way to
+achieve predictable behavior.  Note that this undefined behavior
+cannot occur if the value of `n_meters` of an indexed meter is $2^W$,
+and the type `S` used to construct the meter is `bit<W>`, since the
+index value could never be out of range.
+
+```
+const int METER1_SIZE = 100;
+Meter<bit<7>>((bit<32>) METER1_SIZE, MeterType_t.bytes) meter1;
```

Comment:
Actually, I neglected to to test compiles with latest p4c before committing this.  As of now, "const int METER1_SIZE = 100;" gives an error.

The line "Meter<bit<7>>(100, MeterType_t.bytes) meter1;" gives no error, so neither should replacing the 100 with a METER1_SIZE defined with "#define METER1_SIZE 100".

For now, I will change the example to the version with #define METER1_SIZE 100, and no (bit<32>) cast, but it would be nice here if an extern constructor could take a compile-time constant as an 'int' type parameter, without having to specify a bit width.